File: framework/languages/dotnet/e2e/StoreFailure/Client/Scenarios/SfD2LongOutageRecoveryScenario.cs
// Verifies SF-D2 Long Outage Recovery behavior.
using System.Diagnostics;
using StoreFailure.Client.Support;
using StoreFailure.Shared;
using Zlink.HttpClient;

namespace StoreFailure.Client.Scenarios;

// SF-D2: after an outage long enough to expire every lease, each node
// re-registers itself before trusting lookups, and the disconnect diff
// waits one heartbeat interval — so the surviving peer is never cut,
// while the peer that actually died during the outage is dropped.
internal static class SfD2LongOutageRecoveryScenario
{
    public static async Task RunAsync(
        ClientOptions options,
        ZLinkHttpClient consumer,
        StoreFailureProcessManager processes,
        ManagedProcess providerB)
    {
        await SfProbe.WaitProviderRoutesAsync(
            consumer,
            options.PollingInterval * 4,
            "SF-D2: provider routes were not ready before the store outage.");

        var trafficWindow = options.OwnerLeaseTtl * 2 + options.OwnerLeaseRenewInterval * 4;
        var traffic = Task.Run(() => DriveTolerantRequestsAsync(
            consumer,
            trafficWindow,
            options));

        await processes.PauseStoreAsync();
        try
        {
            // All leases expire during this window, and api-b really dies:
            // recovery has to tell a stale-but-alive peer from a dead one.
            await providerB.KillAsync();
            await Task.Delay(options.OwnerLeaseTtl + options.OwnerLeaseRenewInterval);
        }
        finally
        {
            await processes.UnpauseStoreAsync();
        }

        // The surviving api-a link is never disconnected: requests that
        // round-robined into the killed peer may time out until the dead
        // transport is noticed, but successes must never stop flowing.
        var replies = await traffic;
        ZlinkStreamAssert.Ensure(
            replies.Any(reply => reply.ProviderRid == "api-a"),
            "SF-D2: no request was served by the surviving provider.");

        // api-a re-registers its lease and row before any disconnect diff.
        await SfProbe.WaitPeersAsync(
            consumer,
            SfProbe.PeerRows(options.OwnerLeaseRenewInterval * 6, present: ["api-a"]),
            "SF-D2: the surviving provider did not re-register after recovery.");

        // The dead peer drops out after the one-heartbeat re-registration
        // grace; the survivor keeps serving.
        await SfProbe.WaitPeersAsync(
            consumer,
            SfProbe.PeerRows(options.OwnerLeaseTtl * 2 + options.OwnerLeaseRenewInterval * 4,
                absent: ["api-b"]),
            "SF-D2: the provider that died during the outage was not dropped after recovery.");
        await WaitForDeadProviderDisconnectAsync(consumer, options);

        for (var i = 0; i < 8; i++)
        {
            var reply = await SfProbe.RequestAsync(consumer, $"sf-d2-after-{i}");
            ZlinkStreamAssert.Ensure(
                reply.ProviderRid == "api-a",
                $"SF-D2: post-recovery request {i} was served by '{reply.ProviderRid}'.");
        }

        Console.WriteLine("scenario SF-D2 passed");
    }

    /// <summary>
    /// Steady request cadence where a request that rode the killed peer's
    /// dying transport may fail, but the surviving link must keep serving:
    /// no success gap longer than a couple of seconds is allowed.
    /// </summary>
    private static async Task<List<ProfileRes>> DriveTolerantRequestsAsync(
        ZLinkHttpClient consumer,
        TimeSpan window,
        ClientOptions options)
    {
        var replies = new List<ProfileRes>();
        var elapsed = Stopwatch.StartNew();
        var lastSuccess = TimeSpan.Zero;
        var maxGap = TimeSpan.Zero;
        var index = 0;
        while (elapsed.Elapsed < window)
        {
            try
            {
                var reply = await SfProbe.RequestAsync(consumer, $"sf-d2-{index++}");
                ZlinkStreamAssert.Ensure(
                    reply.Value == "profile:fast",
                    "SF-D2: request returned an unexpected value.");
                replies.Add(reply);
                var gap = elapsed.Elapsed - lastSuccess;
                if (gap > maxGap) maxGap = gap;

                lastSuccess = elapsed.Elapsed;
            }
            catch
            {
                // A request that landed on the killed peer's transport.
            }

            await Task.Delay(150);
        }

        var finalGap = elapsed.Elapsed - lastSuccess;
        if (finalGap > maxGap) maxGap = finalGap;

        ZlinkStreamAssert.Ensure(replies.Count > 0, "SF-D2: the request window produced no successful traffic.");
        ZlinkStreamAssert.Ensure(
            maxGap < options.OwnerLeaseTtl * 2,
            $"SF-D2: successful traffic stalled for {maxGap}, longer than the dead-transport tolerance.");
        return replies;
    }

    private static async Task WaitForDeadProviderDisconnectAsync(
        ZLinkHttpClient consumer,
        ClientOptions options)
    {
        var timeout = options.OwnerLeaseRenewInterval + options.PollingInterval * 4;
        var elapsed = Stopwatch.StartNew();
        var probeTimeoutMilliseconds = (int)Math.Clamp(
            Math.Ceiling(options.PollingInterval.TotalMilliseconds),
            1,
            1000);
        Exception? last = null;
        while (elapsed.Elapsed < timeout)
        {
            var allOnSurvivor = true;
            for (var i = 0; i < 4; i++)
            {
                try
                {
                    var reply = await SfProbe.RequestAsync(
                        consumer,
                        $"sf-d2-disconnect-probe-{Guid.NewGuid():N}",
                        probeTimeoutMilliseconds);
                    if (reply.ProviderRid != "api-a")
                    {
                        allOnSurvivor = false;
                        break;
                    }
                }
                catch (Exception ex)
                {
                    last = ex;
                    allOnSurvivor = false;
                    break;
                }
            }

            if (allOnSurvivor) return;

            await Task.Delay(options.PollingInterval, CancellationToken.None);
        }

        throw new InvalidOperationException(
            "SF-D2: the dead provider transport was still selected after the disconnect grace.",
            last);
    }
}

File: framework/languages/node/e2e/DiscoveryRegistryHa/Client/Scenarios/SfD2LongOutageRecoveryScenario.ts
// SF-D2: 긴 장애 뒤 재등록한 provider만 유지한다 시나리오를 검증한다.
import type { ProfileRes } from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { getJson, getStatus, postJson } from '../../../http-client';
import { ensure } from '../Support/scenario-assert';

interface PeerDto {
  readonly endpoint: string;
  readonly nodeRid?: string;
}

export async function runSfD2(options: ClientOptions): Promise<void> {
  console.log('scenario-control SF-D2 stop-redis-and-kill-api-b');
  await waitForProviderStopped(options.providerBUrl);
  const traffic = driveRequests(options.consumerUrl, 10000);
  console.log('scenario-control SF-D2 restart-redis');
  const replies = await traffic;
  ensure(replies.some((reply) => reply.providerRid === 'api-a'), 'SF-D2 no request was served by surviving api-a.');

  await waitForPeer(options.consumerUrl, 'api-a');
  await waitForMissingPeer(options.consumerUrl, 'api-b');

  for (let i = 0; i < 8; i++) {
    const reply = await postJson<ProfileRes>(options.consumerUrl, '/profile/request', { value: `sf-d2-after-${i}` });
    ensure(reply.value === `profile:sf-d2-after-${i}`, `SF-D2 post-recovery request ${i} value mismatch.`);
    ensure(reply.providerRid === 'api-a', `SF-D2 post-recovery request ${i} was served by '${reply.providerRid}'.`);
  }

  console.log('scenario SF-D2 passed');
}

async function delay(milliseconds: number): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, milliseconds));
}

async function waitForProviderStopped(baseUrl: string): Promise<void> {
  const deadline = Date.now() + 5000;
  while (Date.now() < deadline) {
    const status = await getStatus(`${baseUrl}/health`).catch(() => undefined);
    if (status === undefined || status < 200 || status >= 300) {
      return;
    }
    await delay(50);
  }
  throw new Error('SF-D2 provider api-b did not stop.');
}

async function driveRequests(baseUrl: string, windowMs: number): Promise<ProfileRes[]> {
  const replies: ProfileRes[] = [];
  const deadline = Date.now() + windowMs;
  let index = 0;
  while (Date.now() < deadline) {
    const reply = await postJson<ProfileRes>(baseUrl, '/profile/request', { value: `sf-d2-${index++}` });
    ensure(reply.value.startsWith('profile:sf-d2-'), 'SF-D2 request returned an unexpected value.');
    replies.push(reply);
    await new Promise((resolve) => setTimeout(resolve, 150));
  }

  ensure(replies.length > 0, 'SF-D2 request window produced no successful traffic.');
  return replies;
}

async function waitForPeer(baseUrl: string, rid: string): Promise<void> {
  const deadline = Date.now() + 10000;
  let last: readonly PeerDto[] = [];
  while (Date.now() < deadline) {
    last = await getJson<PeerDto[]>(baseUrl, '/location/peers');
    if (last.some((peer) => peerHasRid(peer, rid))) {
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  throw new Error(`SF-D2 expected live peer ${rid}, last=${JSON.stringify(last)}`);
}

async function waitForMissingPeer(baseUrl: string, rid: string): Promise<void> {
  const deadline = Date.now() + 10000;
  let last: readonly PeerDto[] = [];
  while (Date.now() < deadline) {
    last = await getJson<PeerDto[]>(baseUrl, '/location/peers');
    if (!last.some((peer) => peerHasRid(peer, rid))) {
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  throw new Error(`SF-D2 expected peer ${rid} to be absent, last=${JSON.stringify(last)}`);
}

function peerHasRid(peer: PeerDto, rid: string): boolean {
  return peer.nodeRid === rid;
}
